File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int size = 0;
        ListNode counter = head;
        while (counter != null) {
            size++;
            counter = counter.next;
        }

        int pos = size - n;

        if (pos == 0) {
            return head.next;
        }
    
        ListNode nodeFinder = head;
        // access the desired node
        for (int i = 0; i < pos - 1; i++) {
            nodeFinder = nodeFinder.next;
        }

        nodeFinder.next = nodeFinder.next.next;
        return head;
    }
}

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # I'll assume values are unique
        visited = set()

        curr = head
        while curr:
            visited.add(curr)

            if curr.next and curr.next in visited:
                return True
            
            curr = curr.next

        return False